lib/signalfx/signalflow/channel.rb

```diff
@@ -0,0 +1,135 @@
+# Channel represents a medium through which SignalFlow messages pass.
+# The main method for it is {#each_message}, which is how you get messages from
+# the channel.
+#
+# Channels are for one-time use only.  Once a channel is detached from (either
+# manually or due to the end of a computation) previous messages will be
+# iterable but nothing new will show up.
+class Channel
+  attr_accessor :name
+  attr_accessor :computation
+  attr_accessor :detached
+
+  def initialize(name, detach_cb)
+    @lock = Mutex.new
+    @detach_lock = Mutex.new
+    @cb_lock = Mutex.new
+    @detached = false
+    @name = name
+    @detach_from_transport = detach_cb
+    @cbs = []
+    # Cache every message that comes in on this channel so that we can replay
```

Comment:
The race is if a message comes in on the websocket before the user sets up the iterator.  It has to be stored somewhere and sent to the iteration block when `each_message*` is called or else the message will never be seen. 

It would be simple enough to just delete the contents of `@messages` upon the first replay of them (essentially treating it like a queue like the Py client uses), but I made it support multiple parallel calls to `each_message*`, which proves useful both internally to keep code more decoupled, as well as for users who might want to process the same stream with different logic without having to attach to multiple channels (and double bandwidth) or write a single function to multiplex the logic.  In order to allow additional callbacks to be added and get all of the messages, they have to be stored for the lifetime of the channel.

However, I see your point about unbounded memory consumption.  Do you think it is acceptable to just bound the size of the messages in `@messages` to some reasonable amount (say, 1000) and still be able to keep some replayability of channels?

Reply:
You logic for processing metadata messages is not going to work with that.